Dockerfile

```diff
@@ -1,6 +1,33 @@
-FROM bitnami/nginx:1.18.0
+FROM registry.access.redhat.com/ubi8/nodejs-14 as builder
+
+ENV NODE_OPTIONS=--max-old-space-size=8192
+
+ARG REACT_APP_GIT_SHA
+ENV REACT_APP_GIT_SHA=$REACT_APP_GIT_SHA
+ARG REACT_APP_VERSION
+ENV REACT_APP_VERSION=$REACT_APP_VERSION
+
+COPY --chown=1001:0 / /src/
+RUN chmod 775 /src/
+WORKDIR /src/
+
+RUN npx yarn install
+RUN npx yarn lint
+RUN npx yarn build
```

Comment:
@mhrivnak after taking your branch for a test-drive, and despite you managed to build the app correctly, I'll advise against building the app in a multi-stage fashion for a few reasons.

1. It slows down the build process. 2 images need to be pulled for each release run.
2. Parsing the output log from each step (install/lint/build) is not as transparent as it is while building inside the build machine itself.
3. The build environment is less important because we won't be using node.js after the `build` folder is created. What matters eventually is to have the `build` folder embedded in the container and served via `nginx`.

#### I suggest having something like this:
```
FROM registry.access.redhat.com/ubi8/nginx-120

COPY --chown=1001:0 /build/ "${NGINX_APP_ROOT}/src/"

# persist these on the final image for later inspection
ARG REACT_APP_GIT_SHA
ENV GIT_SHA=$REACT_APP_GIT_SHA
ARG REACT_APP_VERSION
ENV VERSION=$REACT_APP_VERSION

WORKDIR /deploy/
COPY deploy/deploy_config.sh /deploy/
COPY deploy/ui-deployment-template.yaml /deploy/
COPY deploy/nginx.conf /deploy/
COPY deploy/start.sh /deploy/


ENTRYPOINT [ "/deploy/start.sh" ]

```

Reply:
This includes reverting these steps in the release.yaml file.

Reply:
1. I assume devs would still want to build on their local machine. But for releases, does a few extra minutes make a difference? Some projects have multiple Dockerfiles for this reason, so maybe it would be a reasonable compromise to have a `Dockerfile.dev` that uses the `build` dir copied in as today?

2. Can you elaborate? How is it different? I would think that whether it runs inside or outside a container, we should be able to get the same output.

3. The main benefit here of a multistage build is having a reference build environment that we can inspect and control, and is the same no matter where it runs. You have the same exact build environment (and results) when using github actions, your laptop, someone else's laptop, an automated quay.io build, openshift CI, or anywhere else. It's valuable for repeatability/reproducibility, and a "software supply chain" perspective (build tools compiled by Red Hat, vs. by ??? [See page 42 here for more](https://github.com/cncf/tag-security/blob/main/supply-chain-security/supply-chain-security-paper/CNCF_SSCP_v1.pdf)). AFAIK all openshift images that build from source are built using multi-stage builds for this reason. You can also see the other assisted-installer containers get built this way. Of course this image is not currently part of a product, but it is being used by the openshift community, and I think we're wise to follow similar best practices.

From my own perspective, the multi-stage build enabled me to build the project from source without installing any dependencies on my computer. That was helpful, and why I made the change as part of this PR

If this is a blocking issue, I can move the proposal for a multi-stage build to a separate PR. But I do think it's a significant best practice worth following, and it enabled me to build the image without having to worry about setting up a build environment myself.

Reply:
> 1. I assume devs would still want to build on their local machine. But for releases, do a few extra minutes make a difference? Some projects have multiple Dockerfiles for this reason, so maybe it would be a reasonable compromise to have a `Dockerfile.dev` that uses the `build` dir copied in as today?  

"Do a few extra minutes make a difference?" - Yes, especially If the process can take just a few seconds.
"I assume devs would still want to build on their local machine" - Depends on their motivation...
If their motivation is to contribute to this project then they will need to set up their own dev environment anyway (even if that is inside a container). Anyway, offering a container development environment is an interesting and useful topic that I'll be happy to explore but in a different thread.

> 2. Can you elaborate? How is it different? I would think that whether it runs inside or outside a container, we should be able to get the same output.  

Sure. Investigating log failures (manually) is never a pleasant task. For that reason, GH actions parse the output of each build step in a clear way. You can tell easily what output belongs to which step, collapse sections that are not of interest, or expand a particular step section after it has failed to ease the task of troubleshooting the issue. On the other hand, having the build machine just call one build step (docker/podman build...) will defeat the purpose of this feature, you'll end up having a very large output that is not split into sections and makes it harder to investigate.
Since building in the build machine provided by GH and inside a container does produce the same output, I tend to favor the first approach. BTW, I would favor the container approach if I would have to use a shared build machine, that provides no guaranty on whether it has been cleaned up properly after its last use or not. This is not the case.

> 3. The main benefit here of a multistage build is having a reference build environment that we can inspect and control, and is the same no matter where it runs. You have the same exact build environment (and results) when using github actions, your laptop, someone else's laptop, an automated quay.io build, openshift CI, or anywhere else. It's valuable for repeatability/reproducibility, and a "software supply chain" perspective (build tools compiled by Red Hat, vs. by ??? [See page 42 here for more](https://github.com/cncf/tag-security/blob/main/supply-chain-security/supply-chain-security-paper/CNCF_SSCP_v1.pdf)). AFAIK all openshift images that build from source are built using multi-stage builds for this reason. You can also see the other assisted-installer containers get built this way. Of course, this image is not currently part of a product, but it is being used by the openshift community, and I think we're wise to follow similar best practices.  

I understand and second your perspective here. Security-wise I have no objection. Just bear in mind that the first action performed in the release pipeline, the one that checks out the code, is performed by a tool (GH predefined [actions/checkout](https://github.com/actions/checkout)) that is not reviewed by Red Hat. If software sources other than Red Hat's shall not be trusted then this can still manifest into an attack vector in a supply chain attack. Now having considered this scenario, why didn't you rather check out the sources from within the UBI? Which software sources can be trusted and which not? How many of the build tasks can be securely delegated to the build machine provided by GH? If the answer is none, then there is no discussion and we should absolutely do a multi-stage build since there's no other option.

> From my own perspective, the multi-stage build enabled me to build the project from source without installing any dependencies on my computer. That was helpful, and why I made the change as part of this PR

You can still achieve this. Just pull the node.js UBI locally, map a directory between the container and your local filesystem in order to gain access to the build directory after it is created and then perform the exact 3 steps you did in the builder stage.

Reply:
Ok, it sounds like the nice interface of github actions is a big motivator for continuing to use it.

For releases, I still don't understand; it happens once every few days and is automated. Why does every minute or second count?

As for the supply chain aspect, yes of course using a build container does not solve all of the problems. However it is a big step forward in reproducibility. Anyone can easily reproduce the build environment without depending on github's CI.

All that said, we could discuss this for a long time. :) How would you like to move toward a decision? Talk with the team? Save this part for later? Shoot it down for good?

Reply:
Let me take it with the team. Tomorrow we'll settle the discussion.